src/index.js

```diff
@@ -46,15 +46,59 @@ var batch = function(list) {
     return out.concat(
       !item || item === true
         ? 0
-        : typeof item[0] === "function"
+        : isFunc(item[0])
         ? [item]
         : batch(item)
     )
   }, EMPTY_ARR)
 }
 
+var compose = function(a, b) {
+  return a && b && (function(x) { return a(b(x)) }) || a || b
+}
+
+var mapEffects = function(effects, actionMap) {
+  return effects.map(function(fx) {
+    return [ fx[0], fx[1], compose(actionMap, fx[2]) ]
+  })
+}
+
+var makeActionMap = function(extract, merge) {
+  return function actionMap(action) {
+    return function(state, data) {
+      return (function resolve(x, state, data) {
+        return isFunc(x)
+          ? resolve(x(extract(state), data), state)
+          : !isArray(x)
+            ? merge(state, x)
+            : !isArray(x[0]) && !isFunc(x[0])
+              ? [merge(state, x[0])].concat(mapEffects(x.slice(1), actionMap))
+              : resolve(
+                x[0],
+                state,
+                isFunc(x[1]) ? x[1](data) : x[1]
+              )
+      })(action, state, data)
+    }
+  }
+}
+
+var mappedDispatch = function(dispatch, map) {
+  return function(action, props) {
+    dispatch(action, props, map)
+  }
+}
+
+export var map = function(extract, merge, x, actionMap = makeActionMap(extract, merge)) {
+  return isArray(x)
+    ? mapEffects(x, actionMap)
+    : isFunc(x)
+      ? actionMap(x)
+      : (x.actionMap = actionMap, x)
+}
+
```

Comment:
@zaceno Can this live in a separate module? `@hyperapp/x` or whatever?

Reply:
possibly yes. I think we could move out `map` and at least the bulk of `makeActionMap`. But where should I move it? 

Also, if it is going in a separate module, perhaps we should rethink the idea of exporting just a single function?

At least, the current `map` is a little funky since we overload it with different behavior depending on type of argument in, and putting it in a separate module would give us more freedom to export multiple functions for specific purposes.